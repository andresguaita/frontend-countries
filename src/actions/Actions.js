import { types } from "../types/types";
import axios from 'axios'

export const getAllCountries = () =>{
    
    return async (dispatch) =>{
        const resp = await axios('https://countriesapp-backend.herokuapp.com/countries/',
        )
        
        return dispatch({type: types.countriesGetAll, payload: resp.data.countries})

    }
}



export const countriesGetById = (id) =>{

    return async (dispatch) =>{
        try {
            const resp = await axios(`https://countriesapp-backend.herokuapp.com/countries/${id}`)
        
            return dispatch({type: types.countriesGetById, payload: resp.data.country})
        } catch (error) {
            console.log(error)
        }
        

    }

}

export const countriesByName= (name) =>{
    return async (dispatch) =>{
        try {
            const resp = await axios(`https://countriesapp-backend.herokuapp.com/countries/`,{
            params:{
                name
            }
        })
        
        return dispatch({type: types.countriesGetByName, payload: resp.data.country})
        
        } catch (error) {
            dispatch(countryNotFound())
        }     

    }
}

export const countryNotFound = () => ({
    type: types.countriesNotFound
})


export const getAllActivities= () =>{
    return async (dispatch) =>{
        try {
            const resp = await axios(`https://countriesapp-backend.herokuapp.com/activity/`)
        return dispatch({type: types.activityGetAll, payload: resp.data.activities})
        
        } catch (error) {
            console.log(error)
        }     

    }
}


export const newActivity = (name, difficulty,duration,season,countries) =>{

    return async (dispatch) =>{
        try {
            const resp = await axios.post(`https://countriesapp-backend.herokuapp.com/activity`,{
                name, 
                difficulty,
                duration,
                season,
                countries
            })
        
            return dispatch({type: types.activityNew, payload: resp.data.activity})

        } catch (error) {
            console.log(error)
        }
 
    }
}

export const filterByAct = (name) =>(
    {
        type: types.countriesGetByAct,
        payload : name
    }
)

export const filterByContinent = (continent) =>(
    {
        type: types.countriesGetByCont,
        payload : continent
    }
)

export const filterByPopulation = (order) =>(
    {
        type: types.countriesGetByPop,
        payload : order
    }
)


export const filterByLetter = (letter) =>(
    {
        type: types.countriesGetByLetter,
        payload : letter
    }
)

export const countriesDetailEmpty = () =>(
    {
        type: types.countriesDetailEmpty
    }
)
