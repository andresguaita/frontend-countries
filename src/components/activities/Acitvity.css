*, ::after, ::before{
    margin: 0;
    box-sizing: border-box;
}


:root{
    --form-colums:repeat(8,1fr);
    --form-rows: 1fr;
    --clip : polygon(35% 0, 100% 0%, 100% 100%, 0% 100%);

}

.form{
    min-height: 600px;
    height: 100vh;
    display: grid;
    grid-template-columns: var(--form-colums);
    grid-template-rows: (var(--form-rows));
   
}

.form__picture{
    grid-column: 5/-1;
    grid-row: 1/-1;
    display: flex;
    align-items: center;
    position: relative;
    
    
}

.form__picture::before{
    content: '';
    display: block;
    width: 100%;
    height: 100%;
    position: absolute;
    background: #244e92;
    position: absolute;
    top: 0;
    left: 0;
    z-index: -1;
    clip-path: var(--clip);
}

.form__img{
    width: 90%;
}

.form__contain{
    grid-column: 2/5;
    grid-row: 1/-1;
    place-content: center;
    display: grid;
    width: 70%;
    grid-auto-rows: max-content;
    gap: 10px;
    margin-left: 10px;
    margin-top: 10px;
    margin-bottom: 10px;
    -webkit-box-shadow: 1px 3px 19px 5px rgba(0,0,0,0.28); 
    box-shadow: 1px 3px 19px 5px rgba(0,0,0,0.28);
    background: #244e92;
    border-radius: 5px;
}

.form__title{
    color: #fff;
    text-align: center;
    font-size: 3rem;

}

.form__subtitle{
    color: #fff;
    text-align: center;
    
}

.form__title::after{
    content: '';
    display: block;
    width: 80px;
    height: 4px;
    
    margin-top: 10px;
}

.input{
    width: 90%;
    height: 40px;
    outline: none;
    border: 2px solid #244e92;
    border-radius: 5px;
    justify-self: center;
}

.form__submit{
    width: 160px;
    color: #000;
    padding: 15px 20px;
    outline: none;
    border: none;
    border-radius: 6px;
    font-weight: 700;
    align-self: center;
    cursor: pointer;
    justify-self: center;
}

.form__submit:disabled {
    cursor: default;
    opacity: .6;
}



.danger{
    width: 90%;
    height: 40px;
    outline: none;
    border-radius: 5px;
    justify-self: center;
    border: 2px solid red;
}

.danger__text{
    text-align: center;
    font-size: 1.2rem;
    color: white;
    margin: 0;
    background: red;
    margin-left: 5px;
    margin-right: 5px;
    border-radius: 5px;
    padding: 5px;
}

.country__contain{
    display: flex;
    height: 30px;
    flex-direction: row;
    flex-wrap: wrap;
    margin: 0;
    gap: 5px;
    align-items: center;
    justify-content: center;
    background: white;
    margin-left: 20px;
    margin-right: 20px;
    border-radius: 5px;
    
}

.fa-times-circle{
    cursor: pointer;
}


.modal__children{
    display: flex;
    flex-direction: column;
    align-items: center;
}

.modal__children img{
    width: 240px;
    height: 200px;
}

.content__select{
	max-width: 400px;
	position: relative;
    display: inline-flex;
    flex-direction: column;
    justify-self: center;
    gap: 10px;
}
 
.content__select select{
	display: inline-block;
	width: 90%;
	cursor: pointer;
  	padding: 7px 10px;
  	height: 35px;
	border-radius: 0;
	background: #f0f0f0;
	color: black;
	font-size: 1em;
	font-family: 
	'Quicksand', sans-serif;
    align-self: center;
	border:2px solid #244e92;
    border-radius: 12px;
    position: relative;
    transition: all 0.25s ease;
}
 
.Link__back{
    display: flex;
    font-size: 1.2rem;
    width: 60%; 
    background: #244e92;
    text-decoration: none;
    color: white;
    display: flex;
    justify-content: center;
    align-items: center;
    padding: 18px;
    margin: auto;
    border-radius: 5px;
}

.undo__addactivity{
    width: 35px;
    height: 35px;
    margin-right: 5px;
}

.cross__addactivity{
    width: 20px;
    height: 25px;
    cursor: pointer;
}
 
.fa-times-circle{
    color: red;
}
