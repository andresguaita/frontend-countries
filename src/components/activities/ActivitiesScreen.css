.activities__contain{
    display: flex;
    flex-direction: column;
    align-self: center;
    justify-content: center;
    margin-left: 19px;
    border: 1px solid black;
    
    
}

.activities__item{
    width: 100%;
    display: flex;
    flex-direction: row;
    text-align: center;
    gap: 10px;
    text-align: center;
    padding: 10px;
}

.activities__title{

    text-align: center;
    margin-top: 5px;
    margin-bottom: 10px;
}

.activities__notfount{
    font-size: 1.2rem;
    text-align: center;

}
