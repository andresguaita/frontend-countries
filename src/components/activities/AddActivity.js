import React, { useEffect, useState } from 'react'
import { useDispatch, useSelector } from 'react-redux'
import { Link } from 'react-router-dom'
import Swal from 'sweetalert2'
import { getAllCountries, newActivity } from '../../actions/Actions'

import './Acitvity.css'


export const AddActivity = () => {

 
  const dispatch = useDispatch()

  const { countries } = useSelector((state) => state.country)

  countries.sort((a, b) => a.name.localeCompare(b.name))



  useEffect(() => {
    dispatch(getAllCountries())
  }, [dispatch])

  const validate = (input) => {
    let error = {}
    if (!input.name) {
      error.name = 'Name is required'
    }
    else if (!/[a-zA-Z ]{2,254}/.test(input.name)) {
      error.name = 'Name is invalid'
    }

    if (!input.duration) {
      error.duration = 'Duration is required'
    }

    else if ((!/^([1-9]\d*(\.|\,)\d*|0?(\.|\,)\d*[1-9]\d*|[1-9]\d*)$/.test(input.duration))) {
      error.duration = 'Duration is invalid'
    }
    return error

  }

  const [input, setInputChange] = useState({
    name: '',
    difficulty: '',
    duration: '',
    season: '',
    countriesS: []
  })


  const [error, setError] = useState({})

  const arrSeason = ["Spring", "Summer", "Autumn", "Winter"]

  const arrDifficulty = ['1', '2', '3', '4', '5']

  
  const handleSelect = ({ target }) => {
    
    setInputChange({
      ...input,
      countriesS: [...input.countriesS, target.value]
    }
    );
  }

  const handleRemoveSelect = (indexItem) =>{
    setInputChange({
      ...input,
      countriesS: input.countriesS.filter((c, index) => index !== indexItem)
    }
    );
  }


  const handleInputChange = ( e ) => {
    setInputChange({
      ...input,
      [e.target.name]: e.target.value
    })

    let ObjError = validate({ ...input, [e.target.name]: e.target.value })
    setError(ObjError)
  }

  const handleSubmit = (e) => {
    e.preventDefault()

    if(input.name.trim().length>0 || input.duration.trim().length>0){
      dispatch(newActivity(input.name, input.difficulty, input.duration, input.season, input.countriesS))
    setInputChange({
      name: '',
      difficulty: '',
      duration: '',
      season: '',
      countriesS: []
    })
    Swal.fire('Success','Activity has been sucessfully created','success')
    
    }
  }

  


  return (

    <section className="form">
      <figure className='form__picture'>
        <img src='/assets/form.svg' alt='form' className='form__img' />
      </figure>
      <form onSubmit={handleSubmit} className='form__contain animate__animated animate__fadeInLeft'>

        <h2 className='form__title'>Create Activity</h2>
        <h4 className='form__subtitle'>Create Awesome Activity!!!</h4>
        {error.name && (
          <p className="danger__text animate__animated animate__fadeIn"> {error.name}</p>
        )}
        <input placeholder="Name of Activity"
          type="text"
          className={error.name ? 'danger' : 'input'}
          value={input.name}
          name={'name'}
          autoComplete='off'
          onChange={handleInputChange} />

        {error.duration && (
          <p className="danger__text animate__animated animate__fadeIn"> {error.duration}</p>
        )}
        <input
          name={'duration'}
          placeholder='Duration of activity in hours'
          autoComplete='off'
          value={input.duration}
          onChange={handleInputChange}
          className={error.duration ? 'danger' : 'input'}
        />
        
        <div className='content__select'>
        <select
  
          name={'country'}
          onChange={handleSelect}
          required
          >
          <option value="">Country</option>
          {countries?.map(country => (<option value={country.id} key={country.id}>{country.name}</option> ))}

        </select>

        <select
         
          name={'difficulty'}
          onChange={handleInputChange}
          required
          >
          <option value="">Dificulty</option>
          {arrDifficulty.map(difficulty => (<option value={difficulty} key={difficulty}>{difficulty}</option>))}
        </select>

        <select
         
          name={'season'}
          onChange={handleInputChange}
          required
          >
          <option value="">Season</option>
          {arrSeason.map(season => (<option value={season} key={season}>{season}</option>))}
        </select>
        { input.countriesS? <div className='country__contain'>
          {input.countriesS?.map((c,index) => (<span key={c} className='country__select' ><div>{c}<i className="fas fa-times-circle" onClick={() => handleRemoveSelect(index)}></i></div> </span>))}
        </div> : null}
        </div>
       
        <button name="submit" type="submit" className='form__submit' disabled={Object.keys(error).length === 0? false : true}>Submit</button>
        <Link to='/home' className='Link__back'><img src='assets/undo-icon.png' alt='city'className='undo__addactivity'/>  Back to home</Link>
      </form>

    </section>


  )
}
