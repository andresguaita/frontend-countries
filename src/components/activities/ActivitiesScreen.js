import React from 'react';
import { useSelector } from 'react-redux';

import './ActivitiesScreen.css'

export const ActivitiesScreen = () => {

    const { countryDetail } = useSelector(state => state.country)
  return (
            <div className='activities__contain'>
                <h1 className='activities__title'>{`${countryDetail.name} Activities`}</h1>
                {countryDetail.activities? countryDetail.activities.length? countryDetail.activities.map(a=>(
                    <div>
                        <section className='activities__item'>
                        <p>{`Name: ${a.name}`}</p>
                        <p>{`Duration: ${a.duration}`}</p>
                        <p>{`Difficulty:  ${a.difficulty}`}</p>
                        <p>{`Season: ${a.season}`}</p>
                    </section>
                    <hr></hr>
                    </div>
                    
                )): <div>
                    <h1 className='activities__notfount'>Activities not found</h1>
                    </div> : null}
            </div>
  )
};
