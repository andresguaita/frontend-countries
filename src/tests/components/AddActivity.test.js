import React from 'react';
import { mount } from 'enzyme';
import { Provider } from 'react-redux'
import { MemoryRouter } from 'react-router-dom';
import configureStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import '@testing-library/jest-dom';
import { AddActivity } from "../../components/activities/AddActivity";
import { newActivity, getAllCountries } from "../../actions/Actions";



jest.mock("../../actions/Actions", () => ({
    newActivity: jest.fn(),
    getAllCountries: jest.fn()
}))


const middlewares = [thunk];
const mockStore = configureStore(middlewares);

const initState = {
    country : 
   { countries:
      [  {
            "id": "VEN",
            "name": "Venezuela",
            "flag": "https://flagcdn.com/ve.svg",
            "continent": "South America",
            "capital": "Caracas",
            "subregion": "South America",
            "area": 916445,
            "population": 28435943
          }
        ]
    }
};

let store = mockStore(initState);
store.dispatch = jest.fn();

const wrapper = mount( 
    <Provider store={ store }>
        <MemoryRouter>
            <AddActivity /> 
        </MemoryRouter>
    </Provider>

)

describe('Test in <AddActivity/> component', () =>{

    beforeEach(()=> {
        store = mockStore(initState);
        jest.clearAllMocks();
    })

    test('should render the <AddActivity/> correctly', () => {

        expect( wrapper ).toMatchSnapshot();
    });


    test('Should not dispatch newActivity action if the fields are empty', () => {

      
      wrapper.find('form').simulate('submit',{
          preventDefault(){}
      })
      expect (newActivity).not.toHaveBeenCalled()
        ;
    });

    
    
})