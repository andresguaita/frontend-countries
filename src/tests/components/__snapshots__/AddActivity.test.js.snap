// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test in <AddActivity/> component should render the <AddActivity/> correctly 1`] = `
<section
  className="form"
>
  <figure
    className="form__picture"
  >
    <img
      alt="form"
      className="form__img"
      src="/assets/form.svg"
    />
  </figure>
  <form
    className="form__contain"
    onSubmit={[Function]}
  >
    <h2
      className="form__title"
    >
      Create Activity
    </h2>
    <h4
      className="form__subtitle"
    >
      Create Awesome Activity!!!
    </h4>
    <input
      autoComplete="off"
      className="input"
      name="name"
      onChange={[Function]}
      placeholder="Name of Activity"
      type="text"
      value=""
    />
    <input
      autoComplete="off"
      className="input"
      name="duration"
      onChange={[Function]}
      placeholder="Duration of activity in hours"
      value=""
    />
    <div
      className="content__select"
    >
      <select
        name="country"
        onChange={[Function]}
        required={true}
      >
        <option
          value=""
        >
          Country
        </option>
        <option
          key="VEN"
          value="VEN"
        >
          Venezuela
        </option>
      </select>
      <select
        name="difficulty"
        onChange={[Function]}
        required={true}
      >
        <option
          value=""
        >
          Dificulty
        </option>
        <option
          key="1"
          value="1"
        >
          1
        </option>
        <option
          key="2"
          value="2"
        >
          2
        </option>
        <option
          key="3"
          value="3"
        >
          3
        </option>
        <option
          key="4"
          value="4"
        >
          4
        </option>
        <option
          key="5"
          value="5"
        >
          5
        </option>
      </select>
      <select
        name="season"
        onChange={[Function]}
        required={true}
      >
        <option
          value=""
        >
          Season
        </option>
        <option
          key="Spring"
          value="Spring"
        >
          Spring
        </option>
        <option
          key="Summer"
          value="Summer"
        >
          Summer
        </option>
        <option
          key="Autumn"
          value="Autumn"
        >
          Autumn
        </option>
        <option
          key="Winter"
          value="Winter"
        >
          Winter
        </option>
      </select>
      <div
        className="country__contain"
      />
    </div>
    <button
      className="form__submit"
      disabled={false}
      name="submit"
      type="submit"
    >
      Submit
    </button>
    <a
      className="Link__back"
      href="/home"
      onClick={[Function]}
    >
      <img
        alt="city"
        className="undo__addactivity"
        src="assets/undo-icon.png"
      />
        Back to home
    </a>
  </form>
</section>
`;
